open-banking-accelerator/components/com.wso2.openbanking.accelerator.gateway/src/main/java/com/wso2/openbanking/accelerator/gateway/executor/dcr/DCRExecutor.java

```diff
@@ -387,29 +388,30 @@ public void postProcessResponse(OBAPIResponseContext obapiResponseContext) {
                 List<String> subscribedAPIIdList = new ArrayList<>();
                 for (JsonElement subscribedAPI : subscribedAPIsResponse.getAsJsonObject().get("list")
                         .getAsJsonArray()) {
                     String apiId = subscribedAPI.getAsJsonObject().get("apiId").getAsString();
                     subscribedAPIIdList.add(apiId);
                 }
-                //check whether the ssa still contains the roles related to the subscribed APIs
-                List<String> unsubscribedAPIs = getUnAuthorizedAPIs(subscribedAPIsResponse.getAsJsonObject()
-                        .get("list").getAsJsonArray(), regulatoryAPIs, allowedRoles);
-                if (!unsubscribedAPIs.isEmpty()) {
-                    //unsubscribe from the apis
-                    for (String subscriptionId : unsubscribedAPIs) {
-                        if (!callDelete(urlMap.get(GatewayConstants.API_GET_SUBSCRIBED).toString()
-                                .concat("/").concat(subscriptionId), GatewayConstants.BEARER_TAG.concat(token))) {
-                            log.error("Error while unsubscribing from APIs");
-                            //delete SP created to call dev portal REST APIs
-                            callDelete(urlMap.get(GatewayConstants.IAM_DCR_URL).toString().concat("/")
-                                    .concat(clientId), basicAuthHeader);
-                            handleInternalServerError(obapiResponseContext,
-                                    OpenBankingErrorCodes.REGISTRATION_INTERNAL_ERROR);
-                            return;
+
+                String softwareStatement = getSoftwareStatement(createdDCRAppDetails);
+                if (softwareStatement != null) {
+                    List<String> unsubscribedAPIs = getUnAuthorizedAPIs(subscribedAPIsResponse.getAsJsonObject()
+                            .get("list").getAsJsonArray(), regulatoryAPIs, getRolesFromSSA(softwareStatement));
+                    if (!unsubscribedAPIs.isEmpty()) {
+                        //unsubscribe from the apis
+                        for (String subscriptionId : unsubscribedAPIs) {
+                            if (!callDelete(urlMap.get(GatewayConstants.API_GET_SUBSCRIBED).toString()
+                                    .concat("/").concat(subscriptionId), GatewayConstants.BEARER_TAG.concat(token))) {
+                                log.error("Error while unsubscribing from APIs");
+                                //delete SP created to call dev portal REST APIs
+                                callDelete(urlMap.get(GatewayConstants.IAM_DCR_URL).toString().concat("/")
+                                        .concat(clientId), basicAuthHeader);
+                                handleInternalServerError(obapiResponseContext,
+                                        OpenBankingErrorCodes.REGISTRATION_INTERNAL_ERROR);
+                                return;
+                            }
```

Comment:
Shall we consider going with something like below?

```suggestion
                final JsonArray subscribedAPIs = subscribedAPIsResponse.getAsJsonObject()
                            .get("list").getAsJsonArray();

                Optional.of(createdDCRAppDetails.getAsString(OpenBankingConstants.SOFTWARE_STATEMENT))
                    .map(getRolesFromSSA)
                    .map(ssaRoles -> getUnAuthorizedAPIs(subscribedAPIs, regulatoryAPIs, ssaRoles)
                    .flatMap(unAuthorizedApi -> String.format("%s/%s", urlMap.get(GatewayConstants.API_GET_SUBSCRIBED).toString() , unAuthorizedApi)
                    .filter(endpoint -> !callDelete(endpoint, GatewayConstants.BEARER_TAG.concat(token)))
                    .findAny()
                    .ifPresent(endpoint -> {
                        log.error("Error while unsubscribing from API: " + endpint);
                        //delete SP created to call dev portal REST APIs
                        callDelete(String.format("%s/%s", urlMap.get(GatewayConstants.IAM_DCR_URL).toString(), clientId), basicAuthHeader);
                        handleInternalServerError(obapiResponseContext, OpenBankingErrorCodes.REGISTRATION_INTERNAL_ERROR);
                    });
```

Reply:
Resolved in [88d0ac2](https://github.com/wso2/financial-open-banking/pull/62/commits/88d0ac2834e0ceba81ceeca4486dd9ac52acd3cc)